### ISSUE TITLE
Possible to avoid fall damage by sneaking

### ISSUE THREAD

**Issue by jmf** [minetest/minetest#329]
When you drop down from any heigh and land on a slab with shift key pressed you won't take damage.
This also occurs when landing on the lower step of a stair, on the upper step you will take normal damage.
I think it may be for all nodeboxes but i think it will work with all nodeboxes. This may be related to this bug:
https://github.com/celeron55/minetest/issues/127

**Comment by Fixer-007** [comment]
I think you can land on anything with shift and take no damage, have no idea why this 4 year bug is not fixed.

**Comment by C1ffisme** [comment]
I remember using it on the 0gb.us server a LONG time ago. So yes, this bug needs to be fixed.

**Comment by Wuzzy2** [comment]
Note: This also works on basically any nodebox, not just slabs. This means, it also works on snow from Minetest Game, too. This is very serious since snow is now pretty common in Minetest Game.

**Comment by Fixer-007** [comment]
Yeah, this is a problem, also don't forget sneak glitch (dev should decide it is bug or feature?)

**Comment by everamzah** [comment]
Not receiving damage whilst sneaking downwards is a bug; sneaking upwards as per normal is a feature.

**Comment by 0-afflatus** [comment]
Is it really such a huge problem? I think it's a feature, why shouldn't players be able able to make super-hero landings? Don't fix.

**Comment by everamzah** [comment]
>why shouldn't players be able able to make super-hero landings?

Is it sarcasm?  Do you really feel this way?  Is it because sneak jumping is often considered a feature?  That behavior is configurable.  Fall damage isn't.

**Comment by 0-afflatus** [comment]
I'm not being sarcastic or rhetorical, this is an issue tracker.
I genuinely don't understand the issue.

**Comment by everamzah** [comment]
It should, at the very least, be configurable.  Not taking damage from a fall, as far as I am concerned, is unwanted behavior and thus is a bug.

**Comment by MarkuBu** [comment]
>  I genuinely don't understand the issue.

The issue is, that you can cheat. You can jump from every height just by putting some slabs in the landing zone. In every game, on every server, even it is not a superhero game

**Comment by everamzah** [comment]
It doesn't have to be a slab.

**Comment by everamzah** [comment]
https://youtu.be/iNlXsGm5FO0?t=25m10s (processing)

**Comment by Wuzzy2** [comment]
@0-afflatus: Why on Earth would anyone *want* this behaviour? Either you enable fall damage completely or not at all, there should be no “fall damage is enabled but not while you're sneaking” option, as this sounds very silly to me, I don't see an use case and it just overcomplicates Minetest for no good reason.

**Comment by everamzah** [comment]
I know you're not a fan of sneak-jumping, and you made a mod to turn it off, but @Wuzzy2 please understand:  I would not play Minetest without it!

Sometimes a glitch turns into a feature (see the entire history of Quake, for example), and if you don't like it, okay, turn it off.  But believe it or not, it's not easy to avoid fall damage when you sneak, it doesn't always work.  Not on solid blocks, anyway, or at least, not for me.  Point being, it takes some skill.

And that is why someone, like myself, might want this behavior.  Not everybody is the same, remember.

**Comment by Wuzzy2** [comment]
Not easy? It's easy as pie, you just have to aim for the edge. And for nodeboxes like Minetest Game's snow you don't even have to do that, for those things zero skill is required. But difficulty is subjective and not my main reason.

This nodebox-vs-cube discrepance makes this “behaviour” inconsistent and depends on a pretty arbitrary attribute (being a nodebox) of a node in order to work. This is reason enough to get rid of this bug.
From a modding perspective, keeping this bug and pretending it's a feature is a bit insane.
For example: Want to create a block which is proof to sneak-jumping? Sorry, you're out of luck. Want to create a block where sneak-jumping only works at the edge? This will work, but only if it's not a nodebox. Oh, come on! ;-)
Pretending that this bug is a feature is not leading us to anywhere, it just lowers the quality of the Lua API and makes Minetest's already confusing controls (and to some extent, the gameplay, too) more confusing.

I can understand your opinion of wanting some way to avoid fall damage. This is okay with me. What is not okay is doing this by just abusing an old bug without considering consistency, modding, more general use cases and silently accepting other problems which come with this decision.
If you want some “sneaking avoids fall damage” feature, then this bug *still* needs to be fixed because of the other problems.
I propose to fix this bug (what a proposal!). After that, *then* we could maybe think of implementing your desired behaviour as a feature which is first *properly designed* and then *implemented* (without using the buggy code). My minimum criteria for such a feature would be:

* Enables modders to change the fall damage behaviour on a per-node basis (instead of arbitrary criteria like being a nodebox)
* Is not surprising for players (i.e. normal fall damage by default)
* Is properly documented in `lua_api.txt`

IMO this is now going way too off-topic here, opening a new issue about your feature request (!) is probably better.

**Comment by t4im** [comment]
Is this distinct to the already manageable sneak glitch?
```lua
set_physics_override {sneak_glitch = false}
```

**Comment by Wuzzy2** [comment]
Yes IMO. The sneak glitch itself allows you to climb up at insane speeds.
The setting `sneak_glitch` currently *both* disables the sneak glitch and the avoidance of fall damage. Probably the second thing was not intentional and is really just a side effect.

By the way, I wrote an article about sneaking: http://wiki.minetest.net/Sneaking

**Comment by everamzah** [comment]
![4b24b366-4b7a-11e6-9ee6-6642ca67759c](https://cloud.githubusercontent.com/assets/1761623/16955696/9f32f686-4da3-11e6-8d19-9f7f6012cd07.gif)

Can you do it on a solid node, consistently?
https://hybriddog.github.io/hairy-avenger/bpgs/sneakbug.html

Sure, slabs, edges, speed of velocity; these all matter, and would benefit from exploitation and an ability to configure as per preference.

**Comment by rubenwardy** [comment]
Sneaking should not allow you to avoid fall damage by default. And this is definitely a bug. It would be okay if it was possible to make a mod to allow the player to brace themselves and reduce fall damage whilst falling, but it's not okay for this to be here without a mod as a blatant bug.

The sneak elevator is also a definite bug, however for reasons of backwards map compatibility, it should be possible to turn on from a mod. It should not be default imo as it makes no sense as a mechanic.

**Comment by rubenwardy** [comment]
If you want to have fun jumping from things, why not make a subgame that allows you to jump from a height and hit a target. Also, I believe you can use water to break your fall. If this is about annoying fall damage, then maybe use fly or set home to counter this.

**Comment by Fixer-007** [comment]
>Not easy? It's easy as pie, you just have to aim for the edge.

So noob, look this from real PRO (me): https://imgur.com/uYkS3Ko <<< This is landing on FLAT area, it has no plants, I just added decorative rock to aim :}

**Comment by everamzah** [comment]
It isn't about jumping around, it's about fall damage.  Normally it works, but as Fixer points out, it's easy to avoid.

**Comment by 0-afflatus** [comment]
Could this be resolved by disabling sneak_glitch by default, but leaving the setting, so those of us that like it can enable it? I guess I should test it and find out. Sneak ladders do make sense - it's a standard rock climbing technique, just remove every other node - it doesn't need to be more complicated than that. Sure the rate of ascent should be slower.

**Comment by everamzah** [comment]
Not everybody uses it or knows about it, that's what's interesting.  It gives a little reward for those who play long enough to figure it out.  It's enabled by default, you can disable it with a mod, and I think that's correct.

I agree the rate of ascension should be lowered or configurable.

**Comment by Fixer-007** [comment]
Maybe there must be some kind of check... let me explain it:
+ if you fall along the wall that look like this you will manage to catch in that middle space and stop and take no damage (so it needs check two nodes (below side and above side)
```
xxx
xx <
xxx
```
in all other cases it will keep falling and will take damage on all parts of blocks without stops...

**Comment by tobyplowy** [comment]
@Fixer-007 you clearly have no idea what you're doing let me explain how you should do it, first off you need some dank memes second you need to do a 360 no scope and when you land successfully you have to kill yourself while playing the sad violin song, sorry I just had to say it :D

**Comment by tobyplowy** [comment]
I really think this bug should get fixed my frends are actually abusing it on my server it's a bit annoying :D (because they are cheating)

**Comment by paramat** [comment]
Now all nodes, see #3667

**Comment by paramat** [comment]
This issue should now address lack of damage on all nodes, as in #3667

**Comment by Wuzzy2** [comment]
Can you please add this to the 0.4.15 milestone?

**Comment by paramat** [comment]
It may be too late to work on this for release, a solution has not even been thought up.

**Comment by paramat** [comment]
#5327 merged.

**Comment by Fixer-007** [comment]
Unfortunately it is still possible to evade the fall damage even after #5327: https://imgur.com/a/emxdQ

I propose to reopen this issue yet again.

**Comment by Wuzzy2** [comment]
Nooooo! :sob:

Successfully reproduced in 22567d107fffe7a6833b96cc99d531e5303b47dd. :cry:

Here are the exact coordinates you can teleport to which reproduces this at 100% reliability:

    (-349.6960144043, 165.5, 16.808000564576)


How to reproduce:

* Create world
    * Mapgen: v7
    * Seed: 10417323611560738801
* Enter this chat command: `/teleport -349.6960144043 165.5 16.808000564576`
* Hold down sneak until you hit the ground

If you held down sneak, you prevent fall damage. If you didn't, you die. :disappointed:

---

I used this simple mod to get my exact coordinates:

    minetest.register_chatcommand("pos", {
    	privs = {},
    	description = "Prints your coordinates",
    	func = function(playername)
    		local player = minetest.get_player_by_name(playername)
    		minetest.chat_send_player(playername, minetest.pos_to_string(player:getpos()))
    		return true
    	end,
    })

I first experimented a bit as demonstrated in the video by Fixer-007. Then I found the “sweet spot” and immediately entered the `/pos` command.

**Comment by Wuzzy2** [comment]
@sfan5

**Comment by Fixer-007** [comment]
Another ways of evasion, you need to jump down with proper trajectory along side collision border or upper collision borders, example:
![default](https://cloud.githubusercontent.com/assets/16494741/24334831/72a50766-127a-11e7-8e99-a43f297bddb8.png)

**Comment by Fixer-007** [comment]
@SmallJoker found source of the problem, it turns out when we tested the bug, we used teleportation back to point of jump, teleporting and jumping within 3 seconds allows you to evade damage (because of cooldown timer). If you wait longer than 3 sec before jump - you can't evade damage anymore.

**Comment by paramat** [comment]
The other issue of https://github.com/minetest/minetest/issues/329#issuecomment-289311843 above can be fixed like this https://github.com/SmallJoker/minetest/blob/cdf897b/src/localplayer.cpp#L277-L284

**Comment by Fixer-007** [comment]
I'm trying to reproduce landing on edge like on that picture while waiting for 3 seconds for proper testing, will post results later.

**Comment by Fixer-007** [comment]
Update: I've retested this while waiting for 3 seconds cool down and I can't reproduce those bugs I posted here with current code. So I conclude that the reason I evaded the damage was 3 second cool down.

Also, if we read original message:
```
When you drop down from any heigh and land on a slab with shift key pressed you won't take damage.
This also occurs when landing on the lower step of a stair, on the upper step you will take normal damage.
I think it may be for all nodeboxes but i think it will work with all nodeboxes. ```

This part seems to be fixed, so this issue can probably closed for sure.
here is already PR available that deals with 3 second damage cooldown period after teleport.
